libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp

```diff
@@ -23,187 +23,200 @@
+    
     if (_port != NULL) {
+        hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&AP_Frsky_Telem::tick, void));
         // we don't want flow control for either protocol
         _port->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
     }
 }
 
 /*
-  send_frames - sends updates down telemetry link for both DPORT and SPORT protocols
-  should be called by main program at 50hz to allow poll for serial bytes
-  coming from the receiver for the SPort protocol
-*/
-void AP_Frsky_Telem::send_frames(uint8_t control_mode)
+ * init - perform required initialisation
+ * for Plane and Rover
+ */
+void AP_Frsky_Telem::init(const AP_SerialManager &serial_manager, uint8_t *control_mode)
 {
-    // return immediately if not initialised
-    if (!_initialised_uart) {
-        return;
+    // check for protocol configured for a serial port - only the first serial port with one of these protocols will then run (cannot have FrSky on multiple serial ports)
+    if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FrSky_D, 0))) {
+        _protocol = AP_SerialManager::SerialProtocol_FrSky_D; // FrSky D protocol (D-receivers)
+    } else if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FrSky_SPort, 0))) {
+        _protocol = AP_SerialManager::SerialProtocol_FrSky_SPort; // FrSky SPort protocol (X-receivers)
     }
 
-    if (_protocol == FrSkySPORT) {
-        if (!_mode_data_ready) {
-            _mode=control_mode;
-            _mode_data_ready = true;
-        }
-        if (!_baro_data_ready) {
-            calc_baro_alt();
-            _baro_data_ready = true;
-        }
-        if (!_gps_data_ready) {
-            calc_gps_position();
-            _gps_data_ready = true;
-        }
-        if (!_sats_data_ready) {
-            calc_gps_sats();
-            _sats_data_ready = true;
-        }
-        if (!_battery_data_ready) {
-            calc_battery();
-            _battery_data_ready = true;
-        }
-    } else {
-        _mode=control_mode;
-        send_hub_frame();
+    _ap.control_mode = control_mode;
+
+    if (_port != NULL) {
+        hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&AP_Frsky_Telem::tick, void));
+        // we don't want flow control for either protocol
+        _port->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
     }
 }
 
 /*
-  init_uart_for_sport - initialise uart for use by sport
-  this must be called from sport_tick which is called from the 1khz scheduler
-  because the UART begin must be called from the same thread as it is used from
+ * send telemetry data
+ * for FrSky SPort Passthrough (OpenTX) protocol (X-receivers)
  */
-void AP_Frsky_Telem::init_uart_for_sport()
-{
-    // sanity check protocol
-    if (_protocol != FrSkySPORT) {
+void AP_Frsky_Telem::send_SPort_Passthrough(void)
+{
+    static bool chunk_sent = true; // set to false whenever we are not done sending a chunk of mavlink message
```

Comment:
yeah, all static variables need to be moved to the class. This only matters if a given class has multiple instances which this does not but we make it a policy not to do it in case that changes,.. which has happened in the past. Doing it this way consumes same memory and safer